src/compiler/checker.ts

```diff
@@ -11622,12 +11630,16 @@ namespace ts {
         }
 
         // In a typed function call, an argument or substitution expression is contextually typed by the type of the corresponding parameter.
-        function getContextualTypeForArgument(callTarget: CallLikeExpression, arg: Expression): Type {
+        function getContextualTypeForArgument(callTarget: CallLikeExpression, arg: Expression, completion = false): Type {
             const args = getEffectiveCallArguments(callTarget);
             const argIndex = indexOf(args, arg);
             if (argIndex >= 0) {
                 const signature = getResolvedOrAnySignature(callTarget);
-                return getTypeAtPosition(signature, argIndex);
+                const type = getTypeAtPosition(signature, argIndex);
+                if (completion && type && type.flags & TypeFlags.Object && (<ObjectType>type).constraintType) {

```

Comment:
This actually won't work for intersection types or type parameters that are constrained by other type parameters.